File: nbsr/protocol/cbor.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Any

import cbor2

from nbsr.protocol.errors import ProtocolViolation
from nbsr.protocol.registry import ErrorCode


@dataclass(frozen=True, slots=True)
class CborLimits:
    max_total_bytes: int
    max_depth: int
    max_array_items: int
    max_map_pairs: int
    max_text_bytes: int
    max_byte_string_bytes: int

    def __post_init__(self) -> None:
        for value in (
            self.max_total_bytes,
            self.max_depth,
            self.max_array_items,
            self.max_map_pairs,
            self.max_text_bytes,
            self.max_byte_string_bytes,
        ):
            if type(value) is not int or value <= 0:
                raise ValueError("CBOR limits must be positive integers")


DEFAULT_LIMITS = CborLimits(
    max_total_bytes=65_536,
    max_depth=16,
    max_array_items=256,
    max_map_pairs=128,
    max_text_bytes=4_096,
    max_byte_string_bytes=32_768,
)


def _invalid() -> ProtocolViolation:
    return ProtocolViolation(
        ErrorCode.NBSR_E_PROFILE_UNSUPPORTED,
        "Unsupported or non-deterministic CBOR",
    )


def _unsupported() -> ProtocolViolation:
    return ProtocolViolation(
        ErrorCode.NBSR_E_PROFILE_UNSUPPORTED,
        "Unsupported CBOR profile value",
    )


def _over_capacity() -> ProtocolViolation:
    return ProtocolViolation(
        ErrorCode.NBSR_E_OVER_CAPACITY,
        "CBOR resource limit exceeded",
    )


def _encode_head(major: int, argument: int) -> bytes:
    initial = major << 5
    if argument < 24:
        return bytes((initial | argument,))
    if argument <= 0xFF:
        return bytes((initial | 24, argument))
    if argument <= 0xFFFF:
        return bytes((initial | 25,)) + argument.to_bytes(2, "big")
    if argument <= 0xFFFFFFFF:
        return bytes((initial | 26,)) + argument.to_bytes(4, "big")
    if argument <= 0xFFFFFFFFFFFFFFFF:
        return bytes((initial | 27,)) + argument.to_bytes(8, "big")
    raise _unsupported()


def _encode_value(
    value: object,
    limits: CborLimits,
    depth: int,
    active_containers: set[int],
    budget: int,
) -> bytes:
    if depth > limits.max_depth:
        raise _over_capacity()
    if value is None:
        encoded = b"\xf6"
        if len(encoded) > budget:
            raise _over_capacity()
        return encoded
    if value is False:
        encoded = b"\xf4"
        if len(encoded) > budget:
            raise _over_capacity()
        return encoded
    if value is True:
        encoded = b"\xf5"
        if len(encoded) > budget:
            raise _over_capacity()
        return encoded
    if type(value) is int:
        if value >= 0:
            encoded = _encode_head(0, value)
        else:
            encoded = _encode_head(1, -1 - value)
        if len(encoded) > budget:
            raise _over_capacity()
        return encoded
    if isinstance(value, bytes):
        if len(value) > limits.max_byte_string_bytes:
            raise _over_capacity()
        head = _encode_head(2, len(value))
        if len(head) + len(value) > budget:
            raise _over_capacity()
        return head + value
    if isinstance(value, str):
        try:
            encoded = value.encode("utf-8")
        except UnicodeEncodeError as exc:
            raise _unsupported() from exc
        if len(encoded) > limits.max_text_bytes:
            raise _over_capacity()
        head = _encode_head(3, len(encoded))
        if len(head) + len(encoded) > budget:
            raise _over_capacity()
        return head + encoded
    if isinstance(value, (list, tuple)):
        if len(value) > limits.max_array_items:
            raise _over_capacity()
        head = _encode_head(4, len(value))
        if len(head) > budget:
            raise _over_capacity()
        container_id = id(value)
        if container_id in active_containers:
            raise _unsupported()
        active_containers.add(container_id)
        try:
            parts = [head]
            remaining = budget - len(head)
            for item in value:
                encoded_item = _encode_value(
                    item,
                    limits,
                    depth + 1,
                    active_containers,
                    remaining,
                )
                parts.append(encoded_item)
                remaining -= len(encoded_item)
        finally:
            active_containers.remove(container_id)
        return b"".join(parts)
    if isinstance(value, dict):
        if len(value) > limits.max_map_pairs:
            raise _over_capacity()
        head = _encode_head(5, len(value))
        if len(head) > budget:
            raise _over_capacity()
        container_id = id(value)
        if container_id in active_containers:
            raise _unsupported()
        active_containers.add(container_id)
        try:
            remaining = budget - len(head)
            entries: list[tuple[bytes, object]] = []
            for key, item in value.items():
                encoded_key = _encode_value(
                    key,
                    limits,
                    depth + 1,
                    active_containers,
                    remaining,
                )
                entries.append((encoded_key, item))
                remaining -= len(encoded_key)
            entries.sort(key=lambda entry: entry[0])
            if any(previous[0] == current[0] for previous, current in zip(entries, entries[1:], strict=False)):
                raise _invalid()
            parts = [head]
            for encoded_key, item in entries:
                parts.append(encoded_key)
                encoded_item = _encode_value(
                    item,
                    limits,
                    depth + 1,
                    active_containers,
                    remaining,
                )
                parts.append(encoded_item)
                remaining -= len(encoded_item)
        finally:
            active_containers.remove(container_id)
        return b"".join(parts)
    raise _unsupported()


def encode_deterministic(
    value: object,
    limits: CborLimits = DEFAULT_LIMITS,
) -> bytes:
    return _encode_value(value, limits, 1, set(), limits.max_total_bytes)


class _StructuralScanner:
    def __init__(self, data: bytes, limits: CborLimits) -> None:
        self.data = data
        self.limits = limits
        self.offset = 0

    def scan(self) -> None:
        self._scan_item(1)
        if self.offset != len(self.data):
            raise _invalid()

    def _take(self, count: int) -> bytes:
        end = self.offset + count
        if end > len(self.data):
            raise _invalid()
        value = self.data[self.offset : end]
        self.offset = end
        return value

    def _argument(self, additional: int) -> int:
        if additional < 24:
            return additional
        widths = {24: 1, 25: 2, 26: 4, 27: 8}
        width = widths.get(additional)
        if width is None:
            raise _invalid()
        argument = int.from_bytes(self._take(width), "big")
        minimum = {24: 24, 25: 0x100, 26: 0x10000, 27: 0x100000000}[additional]
        if argument < minimum:
            raise _invalid()
        return argument

    def _scan_item(self, depth: int) -> None:
        if depth > self.limits.max_depth:
            raise _over_capacity()
        initial = self._take(1)[0]
        major = initial >> 5
        additional = initial & 0x1F

        if major in (0, 1):
            self._argument(additional)
            return
        if major in (2, 3):
            length = self._argument(additional)
            maximum = self.limits.max_byte_string_bytes if major == 2 else self.limits.max_text_bytes
            if length > maximum:
                raise _over_capacity()
            content = self._take(length)
            if major == 3:
                try:
                    content.decode("utf-8")
                except UnicodeDecodeError as exc:
                    raise _invalid() from exc
            return
        if major == 4:
            item_count = self._argument(additional)
            if item_count > self.limits.max_array_items:
                raise _over_capacity()
            for _ in range(item_count):
                self._scan_item(depth + 1)
            return
        if major == 5:
            pair_count = self._argument(additional)
            if pair_count > self.limits.max_map_pairs:
                raise _over_capacity()
            previous_key: bytes | None = None
            for _ in range(pair_count):
                key_start = self.offset
                self._scan_item(depth + 1)
                encoded_key = self.data[key_start : self.offset]
                if previous_key is not None and encoded_key <= previous_key:
                    raise _invalid()
                previous_key = encoded_key
                self._scan_item(depth + 1)
            return
        if major == 6:
            raise _unsupported()
        if major == 7 and additional in (20, 21, 22):
            return
        raise _unsupported()


def decode_deterministic(
    data: bytes,
    limits: CborLimits = DEFAULT_LIMITS,
) -> Any:
    if not isinstance(data, bytes):
        raise _unsupported()
    if len(data) > limits.max_total_bytes:
        raise _over_capacity()

    _StructuralScanner(data, limits).scan()
    try:
        value = cbor2.loads(data)
    except (cbor2.CBORDecodeError, UnicodeDecodeError, ValueError, TypeError) as exc:
        raise _invalid() from exc
    if encode_deterministic(value, limits) != data:
        raise _invalid()
    return value

File: tools/core-v02-node-verifier/lib/cbor.mjs
import { TextDecoder, TextEncoder } from "node:util";

export const DEFAULT_LIMITS = Object.freeze({
  maxTotalBytes: 65_536,
  maxDepth: 16,
  maxArrayItems: 256,
  maxMapPairs: 128,
  maxTextBytes: 4_096,
  maxByteStringBytes: 32_768,
});

export class CborError extends Error {
  constructor(kind, message) {
    super(message);
    this.name = "CborError";
    this.kind = kind;
  }
}

export class CborTag {
  constructor(tag, value) {
    this.tag = tag;
    this.value = value;
    Object.freeze(this);
  }
}

const UTF8_DECODER = new TextDecoder("utf-8", { fatal: true });
const UTF8_ENCODER = new TextEncoder();
const WIDTHS = new Map([
  [24, 1],
  [25, 2],
  [26, 4],
  [27, 8],
]);
const MINIMUMS = new Map([
  [24, 24n],
  [25, 0x100n],
  [26, 0x10000n],
  [27, 0x100000000n],
]);

function profile(message = "unsupported or non-deterministic CBOR") {
  throw new CborError("profile", message);
}

function capacity(message = "CBOR resource limit exceeded") {
  throw new CborError("capacity", message);
}

function limitsFrom(options) {
  const limits = { ...DEFAULT_LIMITS, ...(options?.limits ?? {}) };
  for (const [name, value] of Object.entries(limits)) {
    if (!Number.isSafeInteger(value) || value < 0) {
      profile(`invalid CBOR limit: ${name}`);
    }
  }
  return limits;
}

function asNumberOrBigInt(value) {
  return value <= BigInt(Number.MAX_SAFE_INTEGER) ? Number(value) : value;
}

class Decoder {
  constructor(data, limits, allowTag18) {
    this.data = data;
    this.limits = limits;
    this.allowTag18 = allowTag18;
    this.offset = 0;
  }

  take(count) {
    const end = this.offset + count;
    if (!Number.isSafeInteger(count) || count < 0 || end > this.data.length) {
      profile("truncated CBOR");
    }
    const value = this.data.subarray(this.offset, end);
    this.offset = end;
    return value;
  }

  argument(additional) {
    if (additional < 24) {
      return BigInt(additional);
    }
    const width = WIDTHS.get(additional);
    if (width === undefined) {
      profile("indefinite or reserved CBOR argument");
    }
    const bytes = this.take(width);
    let value = 0n;
    for (const byte of bytes) {
      value = (value << 8n) | BigInt(byte);
    }
    if (value < MINIMUMS.get(additional)) {
      profile("non-preferred CBOR argument");
    }
    return value;
  }

  count(value, limit) {
    if (value > BigInt(limit) || value > BigInt(Number.MAX_SAFE_INTEGER)) {
      capacity();
    }
    return Number(value);
  }

  item(depth) {
    if (depth > this.limits.maxDepth) {
      capacity();
    }
    const initial = this.take(1)[0];
    const major = initial >> 5;
    const additional = initial & 0x1f;

    if (major === 0) {
      return asNumberOrBigInt(this.argument(additional));
    }
    if (major === 1) {
      const argument = this.argument(additional);
      const value = -1n - argument;
      return value >= BigInt(Number.MIN_SAFE_INTEGER) ? Number(value) : value;
    }
    if (major === 2 || major === 3) {
      const lengthLimit = major === 2
        ? this.limits.maxByteStringBytes
        : this.limits.maxTextBytes;
      const length = this.count(this.argument(additional), lengthLimit);
      const content = this.take(length);
      if (major === 2) {
        return Buffer.from(content);
      }
      try {
        return UTF8_DECODER.decode(content);
      } catch {
        profile("invalid UTF-8 text string");
      }
    }
    if (major === 4) {
      const count = this.count(this.argument(additional), this.limits.maxArrayItems);
      const result = [];
      for (let index = 0; index < count; index += 1) {
        result.push(this.item(depth + 1));
      }
      return result;
    }
    if (major === 5) {
      const count = this.count(this.argument(additional), this.limits.maxMapPairs);
      const result = new Map();
      let previousKey = null;
      for (let index = 0; index < count; index += 1) {
        const keyStart = this.offset;
        const key = this.item(depth + 1);
        const encodedKey = this.data.subarray(keyStart, this.offset);
        if (previousKey !== null && Buffer.compare(previousKey, encodedKey) >= 0) {
          profile("duplicate or incorrectly ordered map key");
        }
        previousKey = encodedKey;
        result.set(key, this.item(depth + 1));
      }
      return result;
    }
    if (major === 6) {
      const tag = this.argument(additional);
      if (!this.allowTag18 || tag !== 18n) {
        profile("unsupported CBOR tag");
      }
      return new CborTag(18, this.item(depth + 1));
    }
    if (major === 7 && additional === 20) {
      return false;
    }
    if (major === 7 && additional === 21) {
      return true;
    }
    if (major === 7 && additional === 22) {
      return null;
    }
    profile("unsupported CBOR simple value or float");
  }
}

function encodeHead(major, argument) {
  if (argument < 0n || argument > 0xffffffffffffffffn) {
    profile("integer outside the supported uint64 range");
  }
  const initial = major << 5;
  if (argument < 24n) {
    return Buffer.from([initial | Number(argument)]);
  }
  for (const [additional, width] of WIDTHS) {
    const maximum = (1n << BigInt(width * 8)) - 1n;
    if (argument <= maximum) {
      const result = Buffer.alloc(1 + width);
      result[0] = initial | additional;
      let remaining = argument;
      for (let index = width; index > 0; index -= 1) {
        result[index] = Number(remaining & 0xffn);
        remaining >>= 8n;
      }
      return result;
    }
  }
  profile("integer outside the supported uint64 range");
}

function checked(parts, budget) {
  const length = parts.reduce((total, part) => total + part.length, 0);
  if (length > budget) {
    capacity();
  }
  return Buffer.concat(parts, length);
}

function encodeValue(value, limits, allowTag18, depth, active, budget) {
  if (depth > limits.maxDepth) {
    capacity();
  }
  if (value === null) {
    return checked([Buffer.from([0xf6])], budget);
  }
  if (value === false) {
    return checked([Buffer.from([0xf4])], budget);
  }
  if (value === true) {
    return checked([Buffer.from([0xf5])], budget);
  }
  if (typeof value === "number" || typeof value === "bigint") {
    if (typeof value === "number" && !Number.isSafeInteger(value)) {
      profile("unsupported numeric value");
    }
    const integer = BigInt(value);
    const encoded = integer >= 0n
      ? encodeHead(0, integer)
      : encodeHead(1, -1n - integer);
    return checked([encoded], budget);
  }
  if (Buffer.isBuffer(value)) {
    if (value.length > limits.maxByteStringBytes) {
      capacity();
    }
    return checked([encodeHead(2, BigInt(value.length)), value], budget);
  }
  if (typeof value === "string") {
    if (!value.isWellFormed()) {
      profile("invalid Unicode string");
    }
    const encoded = Buffer.from(UTF8_ENCODER.encode(value));
    if (encoded.length > limits.maxTextBytes) {
      capacity();
    }
    return checked([encodeHead(3, BigInt(encoded.length)), encoded], budget);
  }
  if (value instanceof CborTag) {
    if (!allowTag18 || value.tag !== 18) {
      profile("unsupported CBOR tag");
    }
    if (active.has(value)) {
      profile("cyclic CBOR value");
    }
    active.add(value);
    try {
      const head = encodeHead(6, 18n);
      const item = encodeValue(
        value.value,
        limits,
        allowTag18,
        depth + 1,
        active,
        budget - head.length,
      );
      return checked([head, item], budget);
    } finally {
      active.delete(value);
    }
  }
  if (Array.isArray(value)) {
    if (value.length > limits.maxArrayItems) {
      capacity();
    }
    if (active.has(value)) {
      profile("cyclic CBOR value");
    }
    active.add(value);
    try {
      const parts = [encodeHead(4, BigInt(value.length))];
      let remaining = budget - parts[0].length;
      for (const item of value) {
        const encoded = encodeValue(item, limits, allowTag18, depth + 1, active, remaining);
        parts.push(encoded);
        remaining -= encoded.length;
      }
      return checked(parts, budget);
    } finally {
      active.delete(value);
    }
  }
  if (value instanceof Map) {
    if (value.size > limits.maxMapPairs) {
      capacity();
    }
    if (active.has(value)) {
      profile("cyclic CBOR value");
    }
    active.add(value);
    try {
      const head = encodeHead(5, BigInt(value.size));
      const entries = [];
      for (const [key, item] of value) {
        entries.push([
          encodeValue(key, limits, allowTag18, depth + 1, active, budget - head.length),
          item,
        ]);
      }
      entries.sort(([left], [right]) => Buffer.compare(left, right));
      for (let index = 1; index < entries.length; index += 1) {
        if (Buffer.compare(entries[index - 1][0], entries[index][0]) === 0) {
          profile("duplicate deterministic map key");
        }
      }
      const parts = [head];
      let remaining = budget - head.length;
      for (const [key, item] of entries) {
        if (key.length > remaining) {
          capacity();
        }
        parts.push(key);
        remaining -= key.length;
        const encoded = encodeValue(item, limits, allowTag18, depth + 1, active, remaining);
        parts.push(encoded);
        remaining -= encoded.length;
      }
      return checked(parts, budget);
    } finally {
      active.delete(value);
    }
  }
  profile("unsupported CBOR value");
}

export function encodeDeterministic(value, options = {}) {
  const limits = limitsFrom(options);
  return encodeValue(
    value,
    limits,
    options.allowTag18 === true,
    1,
    new Set(),
    limits.maxTotalBytes,
  );
}

export function decodeDeterministic(data, options = {}) {
  if (!Buffer.isBuffer(data)) {
    profile("CBOR input must be a Buffer");
  }
  const limits = limitsFrom(options);
  if (data.length > limits.maxTotalBytes) {
    capacity();
  }
  const decoder = new Decoder(data, limits, options.allowTag18 === true);
  const value = decoder.item(1);
  if (decoder.offset !== data.length) {
    profile("trailing CBOR bytes");
  }
  if (!encodeDeterministic(value, { limits, allowTag18: options.allowTag18 === true }).equals(data)) {
    profile("CBOR does not re-encode deterministically");
  }
  return value;
}
